hazelcast/src/main/java/com/hazelcast/map/impl/recordstore/CompositeMutationObserver.java

```diff
@@ -20,234 +20,90 @@
 import com.hazelcast.map.impl.record.Record;
 
 import javax.annotation.Nonnull;
-import java.util.Collection;
-import java.util.LinkedList;
+import java.util.ArrayList;
+import java.util.List;
 
-import static com.hazelcast.internal.util.CollectionUtil.isEmpty;
-import static com.hazelcast.internal.util.ExceptionUtil.rethrow;
+import static java.util.Collections.EMPTY_LIST;
 
 class CompositeMutationObserver<R extends Record> implements MutationObserver<R> {
 
-    private Collection<MutationObserver<R>> mutationObservers;
+    private static final int DEFAULT_OBSERVER_COUNT = 4;
+
+    private List<MutationObserver> mutationObservers = EMPTY_LIST;
 
     CompositeMutationObserver() {
     }
 
-    void add(MutationObserver<R> mutationObserver) {
-        if (mutationObservers == null) {
-            mutationObservers = new LinkedList<>();
+    public void add(MutationObserver<R> mutationObserver) {
+        if (mutationObservers == EMPTY_LIST) {
+            mutationObservers = new ArrayList<>(DEFAULT_OBSERVER_COUNT);
         }
         mutationObservers.add(mutationObserver);
     }
 
     @Override
     public void onClear() {
-        if (isEmpty(mutationObservers)) {
-            return;
-        }
-
-        Throwable exception = null;
-
-        for (MutationObserver<R> mutationObserver : mutationObservers) {
-            try {
-                mutationObserver.onClear();
-            } catch (Throwable t) {
-                if (exception == null) {
-                    exception = t;
-                }
-            }
-        }
-
-        if (exception != null) {
-            rethrow(exception);
+        for (int i = 0; i < mutationObservers.size(); i++) {
+            mutationObservers.get(i).onClear();
```

Comment:
Why are we removing the try-catch wrapping? This allows observers to continue maintaining their internal state even when a previous one misbehaves.

Reply:
Since these observers are part of internal workings of imap(not a user code i mean), proceeding with next observer in case of failure doesn't seem reasonable to me. In case of failure, we should stop there. But i will be happy to hear it, if there is a counter example.

Reply:
@petrpleshachkov wdyt?

Reply:
The idea was to keep indexes up-to date with the data if something goes wrong in one observer. Otherwise, dangling pointers may arise. I'd keep this try-catch.

Reply:
As a rule of thumb, all observers are effectively isolated. If one observer fails due to one other's implementation and we continue proceeding next, this will mean to hide the issue.
Otherwise we can have a situation like this for example, on event journal we see an entry but on index we can't see it, this means we can have different state between observes and this breaks data consistency.